esbulk.go

```diff
@@ -0,0 +1,200 @@
+func ProcessLDJFile(r io.Reader, options Options) (int, error) {
+	count := 0
+
+	if options.Index == "" {
+		return count, errors.New("index name required")
+	}
+
+	if options.Verbose {
+		log.Println(options)
+	}
+
+	if options.Purge {
+		if err := DeleteIndex(options); err != nil {
+			return count, err
+		}
+		time.Sleep(5 * time.Second)
+	}
+
+	// Create index if not exists.
+	if err := CreateIndex(options); err != nil {
+		return count, err
+	}
+
+	if options.Mapping != "" {
+		var reader io.Reader
+		if _, err := os.Stat(options.Mapping); os.IsNotExist(err) {
+			reader = strings.NewReader(options.Mapping)
+		} else {
+			file, err := os.Open(options.Mapping)
+			if err != nil {
+				return count, err
+			}
+			reader = bufio.NewReader(file)
+		}
+
+		if err := PutMapping(options, reader); err != nil {
+			return count, err
+		}
+	}
+
+	queue := make(chan string)
+	var wg sync.WaitGroup
+
+	for i := 0; i < options.NumWorkers; i++ {
+		wg.Add(1)
+		go Worker(fmt.Sprintf("worker-%d", i), options, queue, &wg)
+	}
+
+	for i := range options.Servers {
+		// Store number_of_replicas settings for restoration later.
+		doc, err := GetSettings(i, options)
+		if err != nil {
+			return count, err
+		}
+
+		// TODO(miku): Rework this.
+		numberOfReplicas := doc[options.Index].(map[string]interface{})["settings"].(map[string]interface{})["index"].(map[string]interface{})["number_of_replicas"]
+		if options.Verbose {
+			log.Printf("on shutdown, number_of_replicas will be set back to %s", numberOfReplicas)
+		}
+
+		// Shutdown procedure. TODO(miku): Handle signals, too.
+		defer func() {
+			// Realtime search & reset number of replicas.
+			if _, err := indexSettingsRequest(fmt.Sprintf(`{"index": {"refresh_interval": "1s", "number_of_replicas": %q}}`, numberOfReplicas), options); err != nil {
+				log.Fatal(err)
+			}
+
+			// Persist documents.
+			if FlushIndex(i, options) != nil {
+				log.Fatal(err)
+			}
+		}()
+
+		// Realtime search and reset number of replicas (if specified).
+		var indexRequest = `{"index": {"refresh_interval": "-1"}}`
+		if options.ZeroReplica {
+			indexRequest = `{"index": {"refresh_interval": "-1", "number_of_replicas": 0}}`
+		}
+		resp, err := indexSettingsRequest(indexRequest, options)
+		if err != nil {
+			return count, err
+		}
+		if resp.StatusCode >= 400 {
+			log.Fatal(resp)
+		}
+	}
+
+	reader := bufio.NewReader(r)
+	if options.GZipped {
+		zreader, err := gzip.NewReader(r)
+		if err != nil {
+			return count, err
+		}
+		reader = bufio.NewReader(zreader)
+	}
+
+	for {
+		line, err := reader.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return count, err
+		}
+		line = strings.TrimSpace(line)
+		if len(line) == 0 {
+			continue
+		}
+		queue <- line
+		count++
+	}
+
+	close(queue)
+	wg.Wait()
+
+	return count, nil
+}
+
+// ParseLDJFilename parses filename to get index options for an index insertion
+func ParseLDJFilename(path string, defaults Options) (Options, error) {
```

Comment:
How about `IndexOptionsFromFilepath`? Just as obove, I'd say it's more descriptive